8343791: Socket.connect API should document whether the socket will be closed when hostname resolution fails or another error occurs

vy opened the pull request:
This PR, addressing 8343791, changes `Socket::connect()` methods to close the `Socket` in the event that the connection cannot be established, the timeout expires before the connection is established, or the socket address is unresolved.

`tier3` tests pass against the 9f00f61d3b7fa42a5e23a04f80bb4bb1a2076ef2.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8344527](https://bugs.openjdk.org/browse/JDK-8344527) to be approved

### Issues
 * [JDK-8343791](https://bugs.openjdk.org/browse/JDK-8343791): Socket.connect API should document whether the socket will be closed when hostname resolution fails or another error occurs (**Bug** - P3)
 * [JDK-8344527](https://bugs.openjdk.org/browse/JDK-8344527): Socket.connect API should document whether the socket will be closed when hostname resolution fails or another error occurs (**CSR**)


### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22160/head:pull/22160` \
`$ git checkout pull/22160`

Update a local copy of the PR: \
`$ git checkout pull/22160` \
`$ git pull https://git.openjdk.org/jdk.git pull/22160/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22160`

View PR using the GUI difftool: \
`$ git pr show -t 22160`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22160.diff">https://git.openjdk.org/jdk/pull/22160.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22160#issuecomment-2479739685)
</details>

AlanBateman commented:
/csr

AlanBateman commented:
I've added a lengthy comment to the JBS issue as there some decisions to make on this issue.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -497,13 +497,10 @@ private Socket(SocketAddress address, SocketAddress localAddr, boolean stream)
-                e.addSuppressed(ce);
-            }
-            throw e;
+        } catch (IOException | IllegalArgumentException | SecurityException error) {
```
Naming the exception "error" is a bit strange as this isn't catching Errors. I think the only change need to the Socket constructor is to change line 500 to catch Throwable, leave the try-catch to add the suppressed exception as it's much more readable.

vy replied:
Fixed in 7e633d8d213a950304c550f93a1697c1f2530a05.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -695,15 +703,9 @@ public void connect(SocketAddress endpoint, int timeout) throws IOException {
-                throw new SocketException("Closed by interrupt");
-            }
-            throw e;
+        } catch (IOException error) {
```
At L680 we reject an endpoint that is not an InetSocketAddresss. I think we should follow this with a check to ensure that the address is resolved, that way Socket does all the validation before delegation to the SocketImpl. It means if getImpl().connect(..) throws any exception or error then you can close the Socket, meaning you can catch Throwable.

dfuch replied:
The spec is still not right: we throw SocketException if the socket is already connected, but we do not close the socket in that case.

dfuch replied:
By comparison SocketChannel throws `java.nio.channels.AlreadyConnectedException` which is a subclass of IllegalSstateException (a runtime exception) and leaves the socket open. I am not sure we could change `Socket::connect` to throw a different exception than `SocketException` in that case due to compatibility reasons. So we now have to choose between either:

- modify the spec to explicitely say that the socket remains open if a SocketException is thrown because the socket is already connected
- modify the behavior to also close the socket in that case

Whatever alternative we choose should be also listed in the CSR.

vy replied:
@AlanBateman, how shall we proceed with this one?

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -616,10 +613,15 @@ void setConnected() {
      * @throws  IOException if an error occurs during the connection, the socket
      *          is already connected or the socket is closed
+     * @throws  UnknownHostException if the endpoint address is
+     *          {@linkplain InetSocketAddress#isUnresolved() not resolved},
```
I think better to say "if the endpoint is an unresolved InetSocketAddress".

vy replied:
Fixed in 7e633d8d213a950304c550f93a1697c1f2530a05.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -616,10 +613,15 @@ void setConnected() {
      *        {@code SocketException} with the interrupt status set.
      * </ol>
+     * <p> Besides above noted cases, the socket will also be closed when a
+     * connection attempt fails with an {@link IOException}.
```
I think it would be better to put the new paragraph before the paragraph on interrupt. I think we'll end up with something like " If the endpoint is an unresolved InetSocketAddress, the connection cannot be established, or the timeout expires before the connection is established, then the Socket is closed and IOException is thrown".

vy replied:
Implemented the requested change in 7e633d8d213a950304c550f93a1697c1f2530a05. Note that I excluded the  timeout-related part for `connect(SocketAddress)`, since it doesn't receive timeout as input. Yet, cloned this changes to `connect(SocketAddress,int)` and retained the timeout-related part. Let me know if you want me to add the timeout-related part to the `connect(SocketAddress)` Javadoc too.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -620,6 +621,8 @@ void setConnected() {
      * @throws  IOException if an error occurs during the connection, the socket
      *          is already connected or the socket is closed
+     * @throws  UnknownHostException if the endpoint is an unresolved
+     *          {@link InetSocketAddress}, the socket is closed
```
It might be better to drop ", the socket is closed" from the exception description as it may be mis-read to mean that it gets thrown when the socket is closed.

vy replied:
Fixed in 424fc8ccd3a495a49f859f0cfc563cf463b3efeb.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -650,12 +658,17 @@ public void connect(SocketAddress endpoint) throws IOException {
      *        {@code SocketException} with the interrupt status set.
      * </ol>
+     * <p> Besides above noted cases, the socket will also be closed when a
+     * connection attempt fails with an {@link IOException}.
```
Is this a left-over, I assume it should be removed.

vy replied:
Doh! :facepalm: Fixed in 424fc8ccd3a495a49f859f0cfc563cf463b3efeb.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -650,12 +658,17 @@ public void connect(SocketAddress endpoint) throws IOException {
      *          is already connected or the socket is closed
-     * @throws  SocketTimeoutException if timeout expires before connecting
+     * @throws  SocketTimeoutException if timeout expires before connecting,
+     *          the socket is closed
```
The updated method description makes it clear that the the socket will be closed if the timeout expires before the connection is established, shouldn't need to change the exception description.

vy replied:
Fixed in 424fc8ccd3a495a49f859f0cfc563cf463b3efeb.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -497,13 +497,10 @@ private Socket(SocketAddress address, SocketAddress localAddr, boolean stream)
+        } catch (Throwable throwable) {
+            // `connect()` already closes the socket on failures, yet `bind()` doesn't, and hence:
+            closeQuietly(throwable);
+            throw throwable;
```
Would you mind adding `assert address instanceof InetSocketAddress isa && !isa.isUnresolved();` to the start of this (private) constructor as that will ensure we have the right pre-conditions and avoid further discussion about other socket address types.

AlanBateman replied:
If you take on board the other feedback then you can reduce the scope of the try-catch to just the bind, e.g.
```
        if (localAddr != null) {
            try {
                bind(localAddr);
            } catch (Throwable ex) {
                try {
                    close();
                } catch (Throwable ce) { ex.addSuppressed(ce); }
                throw ex;
            }
        }
        connect(address);   // closes Socket if it throws
```

vy replied:
Fixed in 424fc8ccd3a495a49f859f0cfc563cf463b3efeb.

AlanBateman replied:
I don't mind if we have a broader try-finally but it would be a bug if connect were to throw IAE. All the validation should be done before delegating to the SocketImpl.

vy replied:
Extended the try-finally to cover both `bind()` and `connect()` exceptions in ce568022f9a0fc9c0e9727911e3856048e1a054c.

AlanBateman replied:
Mark Sheppard spotted that too, see exchange in JBS issue. You can drop the !isUnresolved from the assert or eagerly throw UHE if attempting to creating a Socket with an unresolved address, either is fine but probably simple to adjust the assert. The downside is that if someone creates a Socket with an unresolved address then it will create a SocketImpl and attempt to connect before it detects the bad input. It's a corner case so not a big deal.

vy replied:
Yes, I was indeed pinged by @msheppar. (Sorry, I missed the JBS conversation.) Removed the `&& !isa.isUnresolved()` from the assertion in 6b13675de44fd24766d355929f9b695a48cacc71.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -695,15 +708,9 @@ public void connect(SocketAddress endpoint, int timeout) throws IOException {
 
         try {
```
At (old) L680, it checks that endpoint is an InetSocketAddress. We should also check that it resolved, as in:
```
        if (epoint.isUnresolved()) {
            close();
            throw new UnknownHostException(epoint.getHostName() + " is unresolved");
        }
```
The benefit is the the constructor does the validation before is delegates to the SocketImpl. We could change the SocketImpl to throw IAE for this case, it doesn't matter because it will never happen if Socket rejects it.

vy replied:
Implemented in 424fc8ccd3a495a49f859f0cfc563cf463b3efeb, along with a test. Are you sure we want to have a `close()` there instead of `closeQuietly()`?

AlanBateman replied:
I can't think of any cases where close could throw here but it would be better to catch any exception or error from close so that it doesn't override the UHE. A supporting method to catch and add as a suppressed exception is okay but maybe think more about the name as "quietly" it misleading, and "parentError" is a bit strange as it's not an Error.

vy replied:
In a21c5e068555dc99e72d63ab29d3fbe75a543080,

* Renamed `closeQuietly()` to `closeSuppressingExceptions()`
* Replaced the `close()` usage with `closeSuppressingExceptions()` where we throw `UHE` in `connect()`

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -571,6 +569,9 @@ void setConnected() {
      * Connects this socket to the server.
      *
+     * <p> If the connection cannot be established, then the socket is closed
+     * and an {@link IOException} is thrown.
```
What happened to "the endpoint is an unresolved InetSocketAddress"? If part of the change is to close when called with an unresolved address then it needs to be part of the specification.

vy replied:
Fixed in 0761b36297679c9cdea65191f34189b3f51169b0.

AlanBateman commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -649,21 +657,21 @@ public void connect(SocketAddress endpoint, int timeout) throws IOException {
             throw new IllegalArgumentException("Unsupported address type");
 
+        if (epoint.isUnresolved()) {
+            UnknownHostException exception = new UnknownHostException(epoint.getHostName() + " is unresolved");
```
In passing, if you replace this with `var uhe = new UnknownHostException(...)` then the later "throw uhe` will be clearer than `throw exception`.

vy replied:
Fixed in 92ef9ebb8eff8d6972d5363bdce6b88f55c4a7c4.

AlanBateman commented on src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java:
```diff
@@ -564,7 +564,7 @@ protected void connect(SocketAddress remote, int millis) throws IOException {
         InetSocketAddress isa = (InetSocketAddress) remote;
         if (isa.isUnresolved()) {
-            throw new UnknownHostException(isa.getHostName());
+            throw new UnknownHostException("Unresolved address: " + isa.getHostName());
```
The changes means this change isn't needed now. I'm really tempted to say we change the unsupported address type to throw UOE, and the unresolved case to throw IAE, but don't have the complicated this PR any more, plus the SocketImpl API is hugely under specified so might result in more discussion that is worth it.

vy replied:
Reverted in d3568d16cd227940d9254bc4212f332f89243bc2.

AlanBateman commented:
The API + implementation changes have gone through many iterations but I think we've got to a good place now. It should be possible to create or update the CSR.

vy commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -624,7 +633,8 @@ public void connect(SocketAddress endpoint) throws IOException {
      *          is already connected or the socket is closed
-     * @throws  SocketTimeoutException if timeout expires before connecting
+     * @throws  UnknownHostException if the endpoint is an unresolved
+     *          {@link InetSocketAddress}
```
Restored in 7d96bffcb8a93f3730d30f6d0d2edb49d701f465. (I misread what you meant [in this comment](https://github.com/openjdk/jdk/pull/22160#discussion_r1849829296), and hence removed.)

vy commented:
> The API + implementation changes have gone through many iterations but I think we've got to a good place now. It should be possible to create or update the CSR.

@AlanBateman, thank you (and @dfuch, and @jaikiran) for kindly assisting me – I'm grateful that you're bearing with me. I addressed your remarks. I will update the CSR promptly. :saluting_face:

AlanBateman commented on src/java.base/share/classes/sun/nio/ch/NioSocketImpl.java:
```diff
@@ -609,7 +609,7 @@ protected void connect(SocketAddress remote, int millis) throws IOException {
             close();
             if (ioe instanceof InterruptedIOException) {
-                throw ioe;
+                throw new SocketException("Closed by interrupt");
```
SocketTimeoutException is InterruptedIOException so need to distinguish this from the case where a virtual thread is interrupt during connect. The change for NioSocketImpl needs to be like this (once exception handle in switch comes then we can improve this).

```
             }
         } catch (IOException ioe) {
             close();
-            if (ioe instanceof InterruptedIOException) {
+            if (ioe instanceof SocketTimeoutException) {
                 throw ioe;
+            } else if (ioe instanceof InterruptedIOException) {
+                assert Thread.currentThread().isVirtual();
+                throw new SocketException("Closed by interrupt");
             } else {
                 throw SocketExceptions.of(ioe, isa);
             }
```

dfuch reviewed (approved):
OK, this looks good. I believe we are in a good place too.
We throw SocketException and leave the socket opened if the socket is already connected, but the API documentation no longer implies that the socket would also get closed in that case, so we're good.

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,271 @@
+    @ParameterizedTest
+    @MethodSource("ctor_should_close_on_failures")
+    @SuppressWarnings("resource")
+    void ctor_should_close_on_failures(TestCase testCase) throws Throwable {
```
Is the style convention for  Java code  to use camel case for all classes, methods and variables ?

vy replied:
Switched back to camel-case in e37ee840b9681aa017f8653df68c69232a6b6c83. (Snake-case convention for test method names is something I adopted several years ago from other projects, since it was helping with making the failure output more human-readable. But I'd be more than happy to stick to the established convention in the code base.)

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,271 @@
+                // Address and port are mostly ineffective.
+                // They just need to be _valid enough_ to reach to the point where both `SocketImpl#bind()` and `SocketImpl#connect()` are invoked.
+                // Failure will be triggered by the injected `SocketImpl`.
+                new Socket(address, 0xDEAD, address, 0xBEEF);
```
Using new Socket(address, 0xDEAD, address, 0xBEEF);       using hex as port number, a bit too clever!  

I think a constant would be preferable for us simple folk.

private static final int DEADSERVERPORT = 0xDEAD; // or  57005

I think it is preferable for the local port to use 0 i.e. an OS ephemeral allocation, as it is moot in this test context

One observation I’d make on this is that, when scanning the code and not carrying the details of the MockSocketImpl with you, one is inclined to think that this will intermittently lead to BindException and ConnectException as both 0xDEAD and 0xBEEF are in the ephemeral port range

Thus, a general question on the scope of testing. When considering the use of mock object in the call flows. Are these tests sufficient for the testing that the spec changes are upheld by the full default implementation ?

vy replied:
Applied these suggestions in e37ee840b9681aa017f8653df68c69232a6b6c83.

> One observation I’d make on this is that, when scanning the code and not carrying the details of the MockSocketImpl with you, one is inclined to think that this will intermittently lead to BindException and ConnectException as both 0xDEAD and 0xBEEF are in the ephemeral port range

Agreed. That's why I documented the choice of address and port just above the line calling the ctor. The new code you suggested using variables instead of hardcoded constants should further help with this.

> Thus, a general question on the scope of testing. When considering the use of mock object in the call flows. Are these tests sufficient for the testing that the spec changes are upheld by the full default implementation?

That is a very good question! Judging from [the CSR](https://bugs.openjdk.org/browse/JDK-8344527), the specification changes concern:

1. `SocketImpl::connect()`  failures (the connection cannot be established or the timeout expires before the connection is established)
2. `Socket::connect(SocketAddress)` is passed an unresolved address

(1) is verified using `connectShouldCloseOnFailures()`, and (2) is verified using `connectShouldCloseOnUnresolvedAddress()`. If, at a later time, `Socket::connect()` gets modified to **not** use the `SocketImpl`, and invalidate the assumptions held in this test, associated test methods will fail, and force the author to take the CSR requirements into account, AFAICT.

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,271 @@
+
+        private static final String ERROR_MESSAGE = "intentional test failure";
+
+        private static final class ForBindFailure {
```
ForConnectionFailure ForBindfFailure are factories abstraction
Their method are factory methods (the world is addicted to of methods) fabricating a TestCase

ofIOException
ofIllegalArgumentException
ofIOException() 

A simple change to ConnectionFailureFactory and BindFailureFactory
With methods like ioExceptionTestCase illegalArgumentException

Lead to readable code like

TestCase.ConnectionFailureFactory.ioExceptionTestCase()

TestCase.BindFaiureFactory.illegalExceptionTestCase()

vy replied:
Applied your suggestions in e37ee840b9681aa017f8653df68c69232a6b6c83.

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,271 @@
+        });
+
+    }
+
```
static List<TestCase> ctorCloseOnFailuresTestCases() {
         return List.of(
                 TestCase.BindFailureFactory.ioExceptionTestCase(),
                 TestCase.ConnectFailureFactory.ioExceptionTestCase(),
                 TestCase.ConnectFailureFactory.illegalArgumentExceptionTestCase(1));
     }

c.f. name change suggestion below

vy replied:
Applied this suggestion in e37ee840b9681aa017f8653df68c69232a6b6c83.

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,272 @@
+                // Failure will be triggered by the injected `SocketImpl`.
+                InetAddress serverAddress = InetAddress.getLoopbackAddress();
+                int deadServerPort = 0xDEAD;
+                new Socket(serverAddress, deadServerPort, null, 0);
```
new Socket(serverAddress, DEADSERVERPORT, null, 0);       // DEADSERVERPORT declared as a constant above

vy replied:
Fixed in 35d680123f83b97b2a62fe99969aa0c18a75e6b1.

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,272 @@
+class CloseOnFailureTest {
+
+    private static final VarHandle SOCKET_IMPL_FACTORY_HANDLE = createSocketImplFactoryHandle();
+
```
private static final int DEADSERVERPORT = 0xDEAD;

vy replied:
Fixed in 35d680123f83b97b2a62fe99969aa0c18a75e6b1.

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,272 @@
+    private static final class MockSocketImpl extends ThrowingSocketImpl {
+
+        private final AtomicInteger closeInvocationCounter = new AtomicInteger(0);
+
```
replacing Error with Exception in the vriable names allows for easier and quicker scanning of a line     

          private final Exception bindException;

         private final Exception connectException;

         private MockSocketImpl(Exception bindException, Exception connectException) {
             this.bindException = bindException;
             this.connectException = connectException;
         }

vy replied:
Replaced all occurrences of `error` in variable names with `exception` in 35d680123f83b97b2a62fe99969aa0c18a75e6b1.

msheppar commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,272 @@
+    void connectShouldCloseOnUnresolvedAddress() throws IOException {
+        MockSocketImpl socketImpl = new MockSocketImpl(null, null);
+        try (Socket socket = new Socket(socketImpl) {}) {
+            InetSocketAddress address = InetSocketAddress.createUnresolved("no.such.host", 0xBEEF);
```
similar to DEADSERVERPORT

specifying a hex value, albeit a joke value, it is still a valid ephemeral port, and distracts from the main invocation InetAddress.createUnresolved, so consider a constant declaration for the port

private static final int NOSUCHHOSTPORT = 0xBEEF;

then the invocation is
InetAddress.createUnresolved("no.such.host", NOSUCHHOSTPORT);

presenting an unabiquous statement that the InetAddress not meant to be reachable !

vy replied:
Reused the `DEAD_SERVER_PORT` constant here in 35d680123f83b97b2a62fe99969aa0c18a75e6b1.

AlanBateman commented:
I think we should update the socket adaptor at the same time, can you add the following to the change:
```
diff --git a/src/java.base/share/classes/sun/nio/ch/SocketAdaptor.java b/src/java.base/share/classes/sun/nio/ch/SocketAdaptor.java
index cbcfd79378c..191fa177dea 100644
--- a/src/java.base/share/classes/sun/nio/ch/SocketAdaptor.java
+++ b/src/java.base/share/classes/sun/nio/ch/SocketAdaptor.java
@@ -35,6 +35,7 @@
 import java.net.SocketException;
 import java.net.SocketOption;
 import java.net.StandardSocketOptions;
+import java.net.UnknownHostException;
 import java.nio.channels.SocketChannel;
 import java.util.Set;
 
@@ -85,6 +86,14 @@ public void connect(SocketAddress remote) throws IOException {
     public void connect(SocketAddress remote, int timeout) throws IOException {
         if (remote == null)
             throw new IllegalArgumentException("connect: The address can't be null");
+        if (remote instanceof InetSocketAddress isa && isa.isUnresolved()) {
+            if (!sc.isOpen())
+                throw new SocketException("Socket is closed");
+            if (sc.isConnected())
+                throw new SocketException("Already connected");
+            close();
+            throw new UnknownHostException(remote.toString());
+        }
         if (timeout < 0)
             throw new IllegalArgumentException("connect: timeout can't be negative");
         try {
```

vy commented:
> I think we should update the socket adaptor at the same time, can you add the following to the change:

@AlanBateman, applied the patch in 36dd2a154129a97ad59aee82d1a1735a3f0cfbb3. Removed unused

```java
private static InetSocketAddress Net::getLoopbackAddress(int port)
```

too. Shall I also add tests [similar to `CloseOnFailureTest`] for `SocketAdaptor` too?

AlanBateman commented:
> > I think we should update the socket adaptor at the same time, can you add the following to the change:
> 
> @AlanBateman, applied the patch in [36dd2a1](https://github.com/openjdk/jdk/commit/36dd2a154129a97ad59aee82d1a1735a3f0cfbb3). Removed unused
> 
> ```java
> private static InetSocketAddress Net::getLoopbackAddress(int port)
> ```
> 
> too. Shall I also add tests [similar to `CloseOnFailureTest`] for `SocketAdaptor` too?

I see Mark is reviewing the test so I didn't put much time into looking at the test. The main thing is for this test is that it exercises Socket.connect and uses Socket::isClosed to test if the Socket is closed or not after the connect fails. I think we need to test at least the following:

- Socket unbound and unconnected, Socket connect fails, Socket should be closed
- Socket bound and unconnected, Socket connect fails, Socket should be closed
- Socket connected, Socket connect fails, Socket should not be closed
- Socket unbound and unconnected, Socket connect to unresolved address, Socket should be closed
- Socket bound and unconnected, Socket connect to unresolved address, Socket should be closed
- Socket connected, Socket connect connect to unresolved address Socket should not be closed

You shouldn't need a mock SocketImpl. I assume you have this because the test is testing the Socket constructors mostly.

For the socket adapter then you could have the tests be parametrised so they are called with an newly created Socket (unbound/unconnected), that way you can test new Socket() and SocketChannel.open().socket().

AlanBateman commented:
I see Mark is reviewing the test so I didn't put much time into looking at the test. The main thing is for this test is that it exercises Socket.connect and uses Socket::isClosed to test if the Socket is closed or not after the connect fails. I think we need to test at least the following:

Socket unbound and unconnected, Socket connect fails, Socket should be closed
Socket bound and unconnected, Socket connect fails, Socket should be closed
Socket connected, Socket connect fails, Socket should not be closed
Socket unbound and unconnected, Socket connect to unresolved address, Socket should be closed
Socket bound and unconnected, Socket connect to unresolved address, Socket should be closed
Socket connected, Socket connect connect to unresolved address Socket should not be closed
You shouldn't need a mock SocketImpl. I assume you have this because the test is testing the Socket constructors mostly.

For the socket adapter then you could have the tests be parametrised so they are called with an newly created Socket (unbound/unconnected), that way you can test new Socket() and SocketChannel.open().socket().

vy commented on test/jdk/java/net/Socket/CloseOnFailureTest.java:
```diff
@@ -0,0 +1,519 @@
+        void socketShouldNotBeClosedWhenConnectFailsUsingConnectedSocket() throws IOException {
+            SocketAddress serverSocketAddress = serverSocket().getLocalSocketAddress();
+            try (Socket socket = createConnectedSocket(serverSocketAddress)) {
+                assertReconnectFailure(() -> socket.connect(REFUSING_SOCKET_ADDRESS));
```
I needed to delegate the assertion to the implementation because when I call `connect()` on a `Socket` created using `SocketChannel.open(address).socket()` and `new Socket(address, port)`, the former throws an `AlreadyConnectedException`, whereas the latter throws a `SocketException`. Is this discrepancy expected?

dfuch replied:
There are some subtle behaviour discrepancies between plain Socket (created with new Socket()) and the SocketAdaptor (created from SocketChannel.socket()). SocketChannel;;connect will throw AlreadyConnectedException, and so that's what the socket adaptor also does. Ideally we'd want to reduce the distance between plain Socket and the Socket adaptor - but changing behaviours in these APIs is always tricky and a bit risky (and can require a lot of efforts). FWIW - SocketChannel does the right thing - but we're stuck with SocketException in the Socket API due to compatibility reason.

vy replied:
Got it. Thanks for the explanation.

vy commented:
> You shouldn't need a mock SocketImpl. I assume you have this because the test is testing the Socket constructors mostly.

@AlanBateman, existing tests are stressing both ctor **and** `connect()`.

AFAICT, a `SocketImpl` mock is required to verify if the `Socket` ctor closes the `SocketImpl` correctly. I don't know any other way to perform this check.

I also used the `SocketImpl` mock to verify the behavior of `connect()` on

1. `UHE`-causing input
2. Various `SocketImpl`-thrown exceptions

The mock enabled the tests to only focus on what needs to be verified and exclude every other variable.

> I see Mark is reviewing the test so I didn't put much time into looking at the test. The main thing is for this test is that it exercises Socket.connect and uses Socket::isClosed to test if the Socket is closed or not after the connect fails. I think we need to test at least the following:
> 
> * Socket unbound and unconnected, Socket connect fails, Socket should be closed
> * Socket bound and unconnected, Socket connect fails, Socket should be closed
> * Socket connected, Socket connect fails, Socket should not be closed
> * Socket unbound and unconnected, Socket connect to unresolved address, Socket should be closed
> * Socket bound and unconnected, Socket connect to unresolved address, Socket should be closed
> * Socket connected, Socket connect connect to unresolved address Socket should not be closed
> 
> You shouldn't need a mock SocketImpl. I assume you have this because the test is testing the Socket constructors mostly.
> 
> For the socket adapter then you could have the tests be parametrised so they are called with an newly created Socket (unbound/unconnected), that way you can test new Socket() and SocketChannel.open().socket().

I've pushed changes implementing explicit tests for each cited 6 scenarios and using two `Socket` factories:

1. The `Socket` constructor
2. `Socket.open().socket()`

AlanBateman commented:
> > You shouldn't need a mock SocketImpl. I assume you have this because the test is testing the Socket constructors mostly.
> 
> @AlanBateman, existing tests are stressing both ctor **and** `connect()`.

Okay, just a bit surprising to lump them into one test. One thing to consider is one test for the Socket.connect exception scenarios that is focused on checking the connect method throws as expected and the expected Socket state after connect fails. Lots of scenarios to test. A separate test could add to the tests for the Socket constructors, it would be a separate PR if you want.

dfuch commented:
The Mock implementation also has the advantage that you don't need to create a real socket - so you can use any port / address you want since no real socket is bound/connected with those parameters.

vy commented:
> > @AlanBateman, existing tests are stressing both ctor **and** `connect()`.
> 
> Okay, just a bit surprising to lump them into one test. One thing to consider is one test for the Socket.connect exception scenarios that is focused on checking the connect method throws as expected and the expected Socket state after connect fails. Lots of scenarios to test. A separate test could add to the tests for the Socket constructors, it would be a separate PR if you want.

@AlanBateman, 1e3a52c97f3301b1459b03574c25787a282217f6 splits the introduced tests into individual files:

1. `CloseOnNioConnectFailureTest` – tests 6 scenarios you cited using sockets obtained from `SocketChannel.open(address, port).socket()`
2. `CloseOnOioConnectFailureTest` – tests 6 scenarios you cited using sockets obtained from `Socket(address)` constructors
3. `CloseOnConnectFailureMockTest` – using a `SocketImpl` mock, verifies that the socket is closed on `connect()` failures
4. `CloseOnCtorFailureMockTest` – using a `SocketImpl` mock, verifies that the socket is closed on constructor failures

As you noted, the last one (i.e., `CloseOnCtorFailureMockTest`) is not _directly_ related with this ticket. Though I would appreciate it if you can reconsider keeping it, because:

* a `SocketImpl` mock along with test scenarios implemented for `CloseOnConnectFailureMockTest` are reused to do the 80% of the necessary plumbing
* AFAIK, there are no other tests checking the socket is closed on ctor failures

All in all, I am fine with whatever direction you want me to follow – be it keeping `CloseOnCtorFailureMockTest`, or creating a new ticket+PR for it. How shall I proceed?

AlanBateman commented on src/java.base/share/classes/sun/nio/ch/SocketAdaptor.java:
```diff
@@ -85,6 +86,14 @@ public void connect(SocketAddress remote) throws IOException {
+            if (!sc.isOpen())
+                    throw new SocketException("Socket is closed");
+            if (sc.isConnected())
+                    throw new SocketException("Already connected");
```
Would you mind changing this to use 4-space instead of 8-space indent?

vy replied:
Fixed in aaf398f328cfb4771df1a7af22090c5117ec6055.

vy commented:
> I scanned the test update. My personal view is that there way too much here. JDK tests typically only use a small subset of the features that JUnit provides. So very surprised to see the use of extensions in these tests. For tests in this area then the priority should be on having tests that are easy to maintain and diagnose when there are test failures, e.g. think about a test failure in 2 years time where someone has to figure out test base with different subclasses for different APIs. I can imagine someone being puzzled by "Oio" in the name. My guess is that you mean "old I/O" but it looks really odd here.
> 
> Daniel may be able to give you other advice. My view is keep it simple, don't over engineer the tests, and think about diagnosability when tests fail. In this case I think we just need one test ConnectFailTest, or better name, that exercises Socket.connect with the various scenarios. Think CloseOnConnectFailureTestBase without the ceremony. The test methods can be parametized so they execute with sockets created with new Socket() and SocketChannel.open().socket().

@AlanBateman, thanks so much for the thorough review and valuable tips. In 945436425fa953889448ebc4f8b7a80e1d4bd9e1, replaced the existing tests with a single `ConnectFailTest` and tried to be conservative on used JUnit features. Let me know if I captured the design you outlined.

> I'm not sure what to say about the mock SocketImpl and testing the Socket constructor, maybe you can create a test only issue in JBS and contribute it that way. I'd prefer this PR to stay focused on Socket.connect.

Removed tests using mocks in dfb1bb919d81cd07de5da702a91eb3895a7af4e6. I will submit a JBS+PR for these.

dfuch commented on test/jdk/java/net/Socket/ServerSocketTestUtil.java:
```diff
@@ -0,0 +1,83 @@
+
+    static void withEphemeralServerSocket(ThrowingConsumer<ServerSocket> serverSocketConsumer) throws Exception {
+        try (ExecutorService executorService = Executors.newVirtualThreadPerTaskExecutor();
+             ServerSocket serverSocket = new ServerSocket(0)) {
```
I'd suggest making sure the server binds to the loopback address. Using the wildcard address is a know source of instabilities in tests.

```
             new ServerSocket(0, 0, InetAddress.getLoopbackAddress())
```

vy replied:
Thanks for the tip! Fixed in c8bc5a3e4c9c77b4e954019cdd9ecd6095eb1e4b.

AlanBateman commented:
> @AlanBateman, thanks so much for the thorough review and valuable tips. In [9454364](https://github.com/openjdk/jdk/commit/945436425fa953889448ebc4f8b7a80e1d4bd9e1), replaced the existing tests with a single `ConnectFailTest` and tried to be conservative on used JUnit features. Let me know if I captured the design you outlined.

ConnectTestFail could be changed to use parameterized tests with a method source than produces a Socket and SocketChannel.open().socket(). If you had that then it would eliminate the duplucate tests and would allow all the functional interfaces to be removed.

I think it good to find shorter names for some fo the methods and variables as they are way too long in places, e.g. connectedSocketShouldNotBeClosedWhenConnectWithUnresolvedAddressFails.

Did you mean to leave ThrowingSocketImpl in the PR, I thought that was for the other issue. Also are you expecting ServerSocketTestUtil to be used by other tests?

vy commented:
> ConnectTestFail could be changed to use parameterized tests with a method source than produces a Socket and SocketChannel.open().socket(). If you had that then it could eliminate the duplicate tests and would allow most or all of the functional interfaces to be removed.

@AlanBateman, I guess I went too far motivated by your earlier remark on using _"a small subset of the features that JUnit provides"_. Introduced `@ParameterizedTest` back. This did not help much with the custom functional interfaces; I replaced them with their JUnit alternatives.

> I think it good to find shorter names for many of the methods and variables as they are way too long in places, e.g. connectedSocketShouldNotBeClosedWhenConnectWithUnresolvedAddressFails.

Shortened them.

> Did you mean to leave ThrowingSocketImpl in the PR, I thought that was for the other issue.

It was an oversight. Deleted.

> Also are you expecting ServerSocketTestUtil to be used by other tests?

Yes, but we can always do that later. Moved `ServerSocketTestUtil::withEphemeralServerSocket()` into `ConnectFailTest`.

AlanBateman commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,268 @@
+ * @bug 8343791
+ * @summary verifies the socket state after `connect()` failures
+ * @library /test/lib
+ * @run junit/othervm --add-opens java.base/java.net=ALL-UNNAMED ConnectFailTest
```
I assume the --add-opens is left over from a few iteration. For future reference, the @modules tag will do this for you, do need to specify it to the @run tag.

vy replied:
You beat me to it. I was just simplifying this to `junit ConnectFailTest` and saw your review. Fixed.

AlanBateman commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,268 @@
+/*
+ * @test
+ * @bug 8343791
+ * @summary verifies the socket state after `connect()` failures
```
I think you can beef up the @summary a bit because the test checks that Socket.connect throws the expected exception and completes with the Socket in the expected state.

vy replied:
Fixed.

AlanBateman commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,268 @@
+        testConnectedSocket(
+                ConnectFailTest::createConnectedNioSocket,
+                executable -> assertThrows(AlreadyConnectedException.class, executable));
+    }
```
I assume that testConnectedSocket and testConnectedNioSocket could be @ParameterizedTest so it's called with a new created Socket.

AlanBateman replied:
It would be called twice with a new created Socket. Socket.connect is specified to throw IOException if already connected, we shouldn't be throwing (or testing for) AlreadyConnectedException here. It may be that Net.translateToSocketException is missing that mapping (we should check that).

vy replied:
I need to verify the thrown exception in the test:

* If the socket passed created using `Socket::new`, a `SocketException("already connected")` is thrown
* If the socket passed created using `SocketChannel.open().socket()`, an `AlreadyConnectedException` is thrown

AFAICT, the verifier needs to be passed along with the socket, and this implies either separate tests, which is what the current code does, or a separate test method source as in

```java
static List<Arguments> connectedSocketFactoriesAndReconnectFailureVerifiers() {
    return List.of(
            Arguments.of(
                    (Function<SocketAddress, Socket>) ConnectFailTest::createConnectedSocket,
                    (Consumer<Executable>) executable -> {
                        SocketException exception = assertThrows(SocketException.class, executable);
                        assertEquals("already connected", exception.getMessage());
                    }),
            Arguments.of(
                    (Function<SocketAddress, Socket>) ConnectFailTest::createConnectedNioSocket,
                    (Consumer<Executable>) executable -> assertThrows(AlreadyConnectedException.class, executable))
    );
}
```

Am I mistaken?

AlanBateman replied:
I checked the socket adaptor and there is needed a bug here. SocketChannel throws AlreadyConnectedException and the socket adaptor should map that to SocketException("already connected"). It doesn't do that so we should fix it.

vy replied:
I will push an enhancement here. Allow me some time, please.

AlanBateman replied:
> I will push an enhancement here. Allow me some time, please.

That's fine, it seems we have lacked good tests for these exceptional cases so probably not surprising that some issues turn up.

vy replied:
For the record, `Net::translateToSocketException` doesn't map `AlreadyConnectedException`.

vy replied:
> I checked the socket adaptor and there is needed a bug here. SocketChannel throws AlreadyConnectedException and the socket adaptor should map that to SocketException("already connected"). It doesn't do that so we should fix it.

@AlanBateman, I presume you want me to fix that mapping in a separate JBS+PR, right?

AlanBateman replied:
If Daniel is okay with it, then it might be easier to fix it here as part of this PR because the changes to add the mapping are touching exactly the focus of this PR, and the new test is testing the condition that exposes this.

vy replied:
@AlanBateman, I've updated `Net` and `ConnectFailTest`. Would you mind checking them, please? (In the meantime, I will start a `tier1..3` run and update the CSR.)

AlanBateman commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,268 @@
+        try (socket) {
+            assertFalse(socket.isBound());
+            assertFalse(socket.isConnected());
+            assertThrows(IOException.class, () -> socket.connect(UNRESOLVED_ADDRESS));
```
The specified exception for this case is UHE so the assertThrows can be sharper.

vy replied:
Fixed.

AlanBateman commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,268 @@
+             ServerSocket serverSocket = new ServerSocket(0, 0, InetAddress.getLoopbackAddress())) {
+            // Accept connections in the background to avoid blocking the caller
+            executorService.submit(() -> acceptConnections(serverSocket));
+            serverSocketConsumer.accept(serverSocket);
```
I think it just adds complexity to use a virtual thread here.

vy replied:
Replaced with `newSingleThreadExecutor()`.

dfuch commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,268 @@
+        }
+    }
+
+    @SuppressWarnings("resource")
```
What is this SuppressWarnings for?

vy replied:
For IntelliJ's `'SocketChannel' used without 'try'-with-resources statement` warning on `open()` method. Am I supposed to not suppress IDE warnings?

dfuch replied:
Oh - OK - I see. That's fine. We usually don't suppress warnings in tests. Maybe we should.

vy replied:
:+1: Switched to a more fine-grained suppression.

dfuch reviewed (approved):
Looks fine to me! Please make sure to run tier2 again to double check we haven't upset anything.

vy commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    /**
+     * Verifies socket is closed when {@code unboundSocket.connect()} fails.
```
The current Javadoc is as follows:

```
Verifies socket is closed when {@code unboundSocket.connect()} fails.
```

Doesn't this express what you want me to say? How shall I rephrase it?

AlanBateman replied:
No big deal, it just looks a bit because "unboundSocket" is not a parameter or field. Leave it if you want.

vy replied:
Gotcha. Rephrased Javadocs.

vy commented on src/java.base/share/classes/sun/nio/ch/Net.java:
```diff
@@ -166,6 +167,8 @@ else if (x instanceof NotYetConnectedException)
         else if (x instanceof AlreadyBoundException)
             nx = newSocketException("Already bound");
+        else if (x instanceof AlreadyConnectedException)
+            nx = newSocketException("already connected");
```
I deviated from the first-letter-in-upper-case convention deliberately to match the message in `Socket`. Are you sure you want me to replace `already connected` with `Already connected` in `Net`?

AlanBateman replied:
There are 40+ places in Socket that throw a SocketException with a fixed string. It looks like all but two start with an uppercase letter. So maybe we fix Socket.connect instead. I realise this is adding more to this PR but touching older code is always likely to find little oddities like this.

vy replied:
In 840fe7aca9e0b7b35d464fc7706fe39de879143a, capitalized only the `Already connected` message in `Socket` – left the other not-capitalized `SocketException` message (i.e., `factory already defined`) as is. (I have the suspicion that this has triggered some test failures... I am checking them in the meantime.)

vy commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private static void continuouslyAcceptConnections(ServerSocket serverSocket) {
```
That swiftly yielded 10 insertions and 56 deletions! :boom: (Implemented in 33fad083d9b3b4a0ed753e2a9b179e9bbf87387c.)

vy commented:
@AlanBateman, `UHE` message discrepancy between `Socket` (which we introduced in this PR) and `SocketImpl` was causing following tests to fail:

```
javax/xml/jaxp/unittest/common/dtd/DOMTest.java
javax/xml/jaxp/unittest/common/dtd/SAXTest.java
javax/xml/jaxp/unittest/common/catalog/SAXTest.java
javax/xml/jaxp/unittest/common/catalog/DOMTest.java
```

due to `test/jaxp/javax/xml/jaxp/unittest/common/util/TestBase.java:351` checking if the exception message is equal to the host name. I updated the `UHE` message in `Socket` to match the one in `SocketImpl`. I have triggered another `tier1..3` run as we speak.

dfuch commented on src/java.base/share/classes/java/net/Socket.java:
```diff
@@ -661,7 +661,7 @@ public void connect(SocketAddress endpoint, int timeout) throws IOException {
 
         if (epoint.isUnresolved()) {
-            var uhe = new UnknownHostException(epoint.getHostName() + " is unresolved");
+            var uhe = new UnknownHostException(epoint.getHostName());
```
In this instance, I believe reverting to the same message as was thrown before and reconciling Socket with SocketImpl is the right call. The change to the exception message was somewhat gratuitous. Since it makes several tests fail, let's avoid that.

vy replied:
Changed as requested.

AlanBateman commented on test/jdk/java/net/Socket/ConnectFailTest.java:
```diff
@@ -0,0 +1,172 @@
+ * @test
+ * @bug 8343791
+ * @summary verifies that `connect()` failures throw expected exception and leave both `Socket` and the underlying
+ *          `SocketImpl` at the same expected state
```
I think you can drop the mention of the underlying SocketImpl from the summary as the test doesn't do that.

vy replied:
Dropped.

AlanBateman reviewed (approved):
I think this change looks good now. I'm sure the many iterations were frustrating but I think it has all worked out, and the test is significantly simpler and easy to maintain when compared to where it started.

vy commented:
`tier1..3` tests pass – commented on the ticket with a link to the associated the CI run.

Alan's approval is invalidated due to the last `ConnectFailTest` summary change I pushed upon his request. @AlanBateman, @dfuch, I would appreciate it if you can take time for one last (hopefully final) approval, please.

> I think this change looks good now. I'm sure the many iterations were frustrating but I think it has all worked out, and the test is significantly simpler and easy to maintain when compared to where it started.

Indeed the result is objectively way better compared to the initial proposal. :star_struck: Thank you all so much for the guidance. :bow: I'm sorry for the frustration I caused – I will do my best to minimize it with every new PR. :sweat_smile: :innocent:

AlanBateman reviewed (approved):
The CSR should be finalized as that will need to be approved before the change can be integrated.

vy commented:
/integrate

jaikiran commented:
I see that Alan and Daniel have approved the latest state of this PR and the CSR has been approved too. tier1, tier2 and tier3 testing has been reported as successful too. I'll go ahead and sponsor this change now.

jaikiran commented:
/sponsor
